code/modules/paperwork/paper.dm

```diff
@@ -279,212 +306,265 @@
+		return TRUE
 
 	else if(IS_PEN(P))
-		if(icon_state == "scrap")
-			to_chat(usr, "<span class='warning'>\The [src] is too crumpled to write on.</span>")
+		if(is_crumpled)
+			to_chat(user, SPAN_WARNING("\The [src] is too crumpled to write on."))
 			return
 
 		var/obj/item/pen/robopen/RP = P
 		if ( istype(RP) && RP.mode == 2 )
 			RP.RenamePaper(user,src)
 		else
-			var/processed_info_links = user.handle_reading_literacy(user, info_links, length(info))
-			if(processed_info_links)
-				show_browser(user, "<HTML><HEAD><TITLE>[name]</TITLE></HEAD><BODY bgcolor='[color]'>[processed_info_links][stamps]</BODY></HTML>", "window=[name]")
-		return
+			interact(user, readonly = FALSE)
+		return TRUE
 
 	else if(istype(P, /obj/item/stamp) || istype(P, /obj/item/clothing/ring/seal))
-		if((!in_range(src, usr) && loc != user && !( istype(loc, /obj/item/clipboard) ) && loc.loc != user && user.get_active_hand() != P))
-			return
-
-		stamps += (stamps=="" ? "<HR>" : "<BR>") + "<i>This paper has been stamped with the [P.name].</i>"
-
-		var/image/stampoverlay = image('icons/obj/bureaucracy.dmi')
-		var/x
-		var/y
-		if(istype(P, /obj/item/stamp/captain) || istype(P, /obj/item/stamp/boss))
-			x = rand(-2, 0)
-			y = rand(-1, 2)
-		else
-			x = rand(-2, 2)
-			y = rand(-3, 2)
-		offset_x += x
-		offset_y += y
-		stampoverlay.pixel_x = x
-		stampoverlay.pixel_y = y
-
-		if(istype(P, /obj/item/stamp/clown))
-			if(!clown)
-				to_chat(user, "<span class='notice'>You are totally unable to use the stamp. HONK!</span>")
-				return
-
-		if(!ico)
-			ico = new
-		ico += "paper_[P.icon_state]"
-		stampoverlay.icon_state = "paper_[P.icon_state]"
-
-		if(!stamped)
-			stamped = new
-		stamped += P.type
-		overlays += stampoverlay
-
-		playsound(src, 'sound/effects/stamp.ogg', 50, 1)
-		to_chat(user, "<span class='notice'>You stamp the paper with your [P.name].</span>")
+		apply_custom_stamp(
+			image('icons/obj/bureaucracy.dmi', icon_state = "paper_[P.icon_state]", pixel_x = rand(-2, 2), pixel_y = rand(-2, 2)),
+			"with \the [P]")
+		playsound(src, 'sound/effects/stamp.ogg', 50, TRUE)
+		to_chat(user, SPAN_NOTICE("You stamp the paper with your [P.name]."))
+		return TRUE
 
 	else if(istype(P, /obj/item/flame))
 		burnpaper(P, user)
+		return TRUE
 
 	else if(istype(P, /obj/item/paper_bundle))
-		if(!can_bundle())
-			return
-		var/obj/item/paper_bundle/attacking_bundle = P
-		attacking_bundle.insert_sheet_at(user, (attacking_bundle.pages.len)+1, src)
-		attacking_bundle.update_icon()
+		var/obj/item/paper_bundle/B = P
+		B.merge(src, user)
+		return TRUE
+	return ..()
+
+/obj/item/paper/proc/try_bundle_with(var/obj/item/paper/other, var/mob/user)
+	if(!can_bundle_with(other))
+		if(user)
+			to_chat(user, SPAN_WARNING("You can't bundle those!"))
+		return
+
+	var/obj/item/paper_bundle/B = new(loc)
+	if(user)
+		user.unEquip(src,   B)
+		user.unEquip(other, B)
```

Comment:
Thou shalt never invoke `unEquip` without checking the return value.

Reply:
True,. Handling the error condition in this case is a bit annoying though. 
I'll just pop a generic warning that you can't drop a paper or something I guess.

Reply:
There's a `canUnEquip` if you need to check it before making the bundle.